kv/kvserver: TestStoreRangeMergeDuringShutdown failed

[(kv/kvserver).TestStoreRangeMergeDuringShutdown failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2446890&tab=buildLog) on [master@e7e0f266226b9c0b664bc35d29eb312efd7d0b37](https://github.com/cockroachdb/cockroach/commits/e7e0f266226b9c0b664bc35d29eb312efd7d0b37):

```
=== RUN   TestStoreRangeMergeDuringShutdown
    test_log_scope.go:155: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreRangeMergeDuringShutdown799109238
    test_log_scope.go:64: use -show-logs to present logs inline
    stopper.go:98: leaked stopper, created at:
        goroutine 160923 [running]:
        runtime/debug.Stack(0xa7d3b00, 0xc00009ec00, 0x0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/util/stop.register(0xc03b48cbe0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:52 +0x45
        github.com/cockroachdb/cockroach/pkg/util/stop.NewStopper(0x0, 0x0, 0x0, 0xc01f221440)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0x39c
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).addStore(0xc01ac8e540, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:883 +0x3d4
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).Start(0xc01ac8e540, 0x8453520, 0xc01d9ed800, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:409 +0xfe5
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeMergeDuringShutdown(0xc01d9ed800)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_merge_test.go:3403 +0x55b
        testing.tRunner(0xc01d9ed800, 0x6ef7650)
        	/usr/local/go/src/testing/testing.go:1123 +0x203
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x5bc
--- FAIL: TestStoreRangeMergeDuringShutdown (0.23s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeMergeDuringShutdown PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeMergeDuringShutdown.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreRangeMergeDuringShutdown failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2447089&tab=buildLog) on [master@7ca612a9f99ecfea2052c71248ec292fbee74775](https://github.com/cockroachdb/cockroach/commits/7ca612a9f99ecfea2052c71248ec292fbee74775):

```
=== RUN   TestStoreRangeMergeDuringShutdown
    test_log_scope.go:155: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreRangeMergeDuringShutdown882537650
    test_log_scope.go:64: use -show-logs to present logs inline
    stopper.go:98: leaked stopper, created at:
        goroutine 160688 [running]:
        runtime/debug.Stack(0xa7d4b40, 0xc000170400, 0x0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/util/stop.register(0xc038f6c5a0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:52 +0x45
        github.com/cockroachdb/cockroach/pkg/util/stop.NewStopper(0x0, 0x0, 0x0, 0xc012a64120)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0x39c
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).addStore(0xc035098e00, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:883 +0x3d4
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).Start(0xc035098e00, 0x8453d00, 0xc020d51b00, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:409 +0xfe5
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeMergeDuringShutdown(0xc020d51b00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_merge_test.go:3403 +0x55b
        testing.tRunner(0xc020d51b00, 0x6ef7d68)
        	/usr/local/go/src/testing/testing.go:1123 +0x203
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x5bc
--- FAIL: TestStoreRangeMergeDuringShutdown (0.26s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeMergeDuringShutdown PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeMergeDuringShutdown.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreRangeMergeDuringShutdown failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2447173&tab=buildLog) on [master@5403e542f095bd93f5c805a3ac73f3ebe9c82b4b](https://github.com/cockroachdb/cockroach/commits/5403e542f095bd93f5c805a3ac73f3ebe9c82b4b):

```
=== RUN   TestStoreRangeMergeDuringShutdown
    test_log_scope.go:155: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreRangeMergeDuringShutdown904994769
    test_log_scope.go:64: use -show-logs to present logs inline
    stopper.go:98: leaked stopper, created at:
        goroutine 160250 [running]:
        runtime/debug.Stack(0xa7d3b00, 0xc000500400, 0x0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/util/stop.register(0xc01a6b3cc0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:52 +0x45
        github.com/cockroachdb/cockroach/pkg/util/stop.NewStopper(0x0, 0x0, 0x0, 0xc01fa34d80)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0x39c
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).addStore(0xc03dadf6c0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:883 +0x3d4
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).Start(0xc03dadf6c0, 0x8453520, 0xc04077aa80, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:409 +0xfe5
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeMergeDuringShutdown(0xc04077aa80)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_merge_test.go:3403 +0x55b
        testing.tRunner(0xc04077aa80, 0x6ef7650)
        	/usr/local/go/src/testing/testing.go:1123 +0x203
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x5bc
--- FAIL: TestStoreRangeMergeDuringShutdown (0.24s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeMergeDuringShutdown PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeMergeDuringShutdown.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreRangeMergeDuringShutdown failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2447194&tab=buildLog) on [master@4e6e7450afa88c877ad28ffe18e96e5c685f1065](https://github.com/cockroachdb/cockroach/commits/4e6e7450afa88c877ad28ffe18e96e5c685f1065):

```
=== RUN   TestStoreRangeMergeDuringShutdown
    test_log_scope.go:155: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreRangeMergeDuringShutdown719373315
    test_log_scope.go:64: use -show-logs to present logs inline
    stopper.go:98: leaked stopper, created at:
        goroutine 163020 [running]:
        runtime/debug.Stack(0xa7d4b40, 0xa7da7a0, 0x0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/util/stop.register(0xc01133f720)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:52 +0x45
        github.com/cockroachdb/cockroach/pkg/util/stop.NewStopper(0x0, 0x0, 0x0, 0xc0348c9560)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0x39c
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).addStore(0xc0248a6a80, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:883 +0x3d4
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).Start(0xc0248a6a80, 0x8453d00, 0xc009096000, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:409 +0xfe5
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeMergeDuringShutdown(0xc009096000)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_merge_test.go:3403 +0x55b
        testing.tRunner(0xc009096000, 0x6ef7d68)
        	/usr/local/go/src/testing/testing.go:1123 +0x203
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x5bc
--- FAIL: TestStoreRangeMergeDuringShutdown (0.25s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeMergeDuringShutdown PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeMergeDuringShutdown.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestStoreRangeMergeDuringShutdown failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2447382&tab=buildLog) on [master@4e6e7450afa88c877ad28ffe18e96e5c685f1065](https://github.com/cockroachdb/cockroach/commits/4e6e7450afa88c877ad28ffe18e96e5c685f1065):

```
=== RUN   TestStoreRangeMergeDuringShutdown
    test_log_scope.go:155: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestStoreRangeMergeDuringShutdown106563231
    test_log_scope.go:64: use -show-logs to present logs inline
    stopper.go:98: leaked stopper, created at:
        goroutine 161094 [running]:
        runtime/debug.Stack(0xa7d4b40, 0xc000488000, 0x0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/util/stop.register(0xc036db90e0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:52 +0x45
        github.com/cockroachdb/cockroach/pkg/util/stop.NewStopper(0x0, 0x0, 0x0, 0xc027a93d40)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0x39c
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).addStore(0xc022e5dc00, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:883 +0x3d4
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.(*multiTestContext).Start(0xc022e5dc00, 0x8453d00, 0xc031b1dc80, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_test.go:409 +0xfe5
        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestStoreRangeMergeDuringShutdown(0xc031b1dc80)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_merge_test.go:3403 +0x55b
        testing.tRunner(0xc031b1dc80, 0x6ef7d68)
        	/usr/local/go/src/testing/testing.go:1123 +0x203
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x5bc
--- FAIL: TestStoreRangeMergeDuringShutdown (0.23s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStoreRangeMergeDuringShutdown PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStoreRangeMergeDuringShutdown.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/util/stop/stopper.go
+++ b/pkg/util/stop/stopper.go
@@ -480,8 +480,11 @@ func (s *Stopper) Stop(ctx context.Context) {
 		return
 	}
 
-	defer s.Recover(ctx)
-	defer unregister(s)
+	defer func() {
+		s.Recover(ctx)
+		unregister(s)
+		close(s.stopped)
+	}()
 
 	if log.V(1) {
 		file, line, _ := caller.Lookup(1)
@@ -495,7 +498,6 @@ func (s *Stopper) Stop(ctx context.Context) {
 	if r := recover(); r != nil {
 		go s.Quiesce(ctx)
 		close(s.stopper)
-		close(s.stopped)
 		s.mu.Lock()
 		for _, c := range s.mu.closers {
 			go c.Close()
@@ -518,7 +520,6 @@ func (s *Stopper) Stop(ctx context.Context) {
 	for _, c := range s.mu.closers {
 		c.Close()
 	}
-	close(s.stopped)
 }
 
 // ShouldQuiesce returns a channel which will be closed when Stop() has been

--- a/pkg/util/stop/stopper_test.go
+++ b/pkg/util/stop/stopper_test.go
@@ -135,8 +135,10 @@ func TestStopperMultipleStopees(t *testing.T) {
 
 func TestStopperStartFinishTasks(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	s := stop.NewStopper()
+
 	ctx := context.Background()
+	s := stop.NewStopper()
+	defer s.Stop(ctx)
 
 	if err := s.RunTask(ctx, "test", func(ctx context.Context) {
 		go s.Stop(ctx)

